zerver/lib/thumbnail.py

```diff
@@ -39,122 +40,106 @@ def generate_thumbnail_url(path: str, size: str = "0x0") -> str:
+    except pyvips.Error:
         raise BadImageError(_("Could not decode image; did you upload an image file?"))
-    except DecompressionBombError:
-        raise BadImageError(_("Image size exceeds limit."))
-    out = io.BytesIO()
-    if im.mode == "CMYK":
-        im = im.convert("RGB")
-    im.save(out, format="png")
-    return out.getvalue()
 
+    if source_image.width * source_image.height > IMAGE_BOMB_TOTAL_PIXELS:
+        raise BadImageError(_("Image size exceeds limit."))
 
-def resize_logo(image_data: bytes) -> bytes:
     try:
-        im = Image.open(io.BytesIO(image_data))
-        im = ImageOps.exif_transpose(im)
-        im.thumbnail((8 * DEFAULT_AVATAR_SIZE, DEFAULT_AVATAR_SIZE), Image.Resampling.LANCZOS)
-    except OSError:
-        raise BadImageError(_("Could not decode image; did you upload an image file?"))
-    except DecompressionBombError:
-        raise BadImageError(_("Image size exceeds limit."))
-    out = io.BytesIO()
-    if im.mode == "CMYK":
-        im = im.convert("RGB")
-    im.save(out, format="png")
-    return out.getvalue()
-
-
-def resize_animated(im: Image.Image, size: int = DEFAULT_EMOJI_SIZE) -> bytes:
-    assert im.n_frames > 1
-    frames = []
-    duration_info = []
-    disposals = []
-    # If 'loop' info is not set then loop for infinite number of times.
-    loop = im.info.get("loop", 0)
-    for frame_num in range(im.n_frames):
-        im.seek(frame_num)
-        new_frame = im.copy()
-        new_frame.paste(im, (0, 0), im.convert("RGBA"))
-        new_frame = ImageOps.pad(new_frame, (size, size), Image.Resampling.LANCZOS)
-        frames.append(new_frame)
-        if im.info.get("duration") is None:  # nocoverage
-            raise BadImageError(_("Corrupt animated image."))
-        duration_info.append(im.info["duration"])
-        if isinstance(im, GifImagePlugin.GifImageFile):
-            disposals.append(
-                im.disposal_method  # type: ignore[attr-defined]  # private member missing from stubs
-            )
-        elif isinstance(im, PngImagePlugin.PngImageFile):
-            disposals.append(im.info.get("disposal", PngImagePlugin.Disposal.OP_NONE))
-        else:  # nocoverage
-            raise BadImageError(_("Unknown animated image format."))
-    out = io.BytesIO()
-    frames[0].save(
-        out,
-        save_all=True,
-        optimize=False,
-        format=im.format,
-        append_images=frames[1:],
-        duration=duration_info,
-        disposal=disposals,
-        loop=loop,
-    )
+        yield source_image
+    except pyvips.Error as e:
+        logging.exception(e)
+        raise BadImageError(_("Bad image!"))
+
+
+def resize_avatar(image_data: bytes, size: int = DEFAULT_AVATAR_SIZE) -> bytes:
+    # This will scale up, if necessary, and will scale the smallest
+    # dimension to fit.  That is, a 1x1000 image will end up with the
+    # one middle pixel enlarged to fill the full square.
+    with libvips_check_image(image_data):
```

Comment:
This seems to not use the thing yielded by `with libvips_check_image`?

Reply:
Some callsites do, some don't.  We're using the contextmanager to verify that the thing parses as an image.  One interesting property of libvips is that it calling `thumbnail_image` on a pre-created `pyvips.Image` is probably much _slower_ than calling `pyvips.Image.thumbnail_buffer()`, because in the latter case it can make use of things like shrink-on-load using JPEG blocks.

I'll add some comments.